tests/contrib/funsor/test_valid_models_enum.py

```diff
@@ -0,0 +1,444 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import defaultdict
+import contextlib
+import logging
+import os
+from queue import LifoQueue
+
+import pytest
+import torch
+
+from pyro.infer.enum import iter_discrete_escape, iter_discrete_extend
+from pyro.ops.indexing import Vindex
+from pyro.poutine import Trace
+from pyro.poutine.util import prune_subsample_sites
+from pyro.util import check_traceenum_requirements
+
+# put all funsor-related imports here, so test collection works without funsor
+try:
+    import funsor
+    import pyro.contrib.funsor
+    from pyro.contrib.funsor.handlers.runtime import _DIM_STACK
+    funsor.set_backend("torch")
+    from pyroapi import distributions as dist
+    from pyroapi import infer, handlers, pyro, pyro_backend
+except ImportError:
+    pytestmark = pytest.mark.skip(reason="funsor is not installed")
+
+logger = logging.getLogger(__name__)
+
+# default to 2, which checks that packed but not unpacked shapes match
+_NAMED_TEST_STRENGTH = int(os.environ.get("NAMED_TEST_STRENGTH", 2))
+
+
+def assert_ok(model, guide=None, max_plate_nesting=None, **kwargs):
```

Comment:
The actual tests in this file are copied verbatim from `tests/infer/test_valid_models.py`, but this function (`assert_ok`) is different and should be reviewed a bit more closely since I use it heavily here and in subsequent PRs.